lib/awful/rules.lua

```diff
@@ -342,10 +342,41 @@ end
 
 rules.add_rule_source("awful.rules", apply_awful_rules, {"awful.spawn"}, {})
 
+local blacklisted_snid = {}
+
+--- Make startup notification work for some clients like XTerm. This is ugly
+-- but works often enough to be useful.
+local function get_startup_id_fallback(c)
+    if not c.pid then return end
+
+    local f = io.open("/proc/"..c.pid.."/environ", "rb")
+
+    -- It will only work on Linux, that's already 99% of the userbase.
+    if not f then return end
+
+    local s = f:read("*all")
+
+    f:close()
+
+    local snid = s:match("STARTUP_ID=(.*)")
```

Comment:
Sigh, this does not work since you also capture all following variables:
```
$ lua -e 'local c = "SOMETHING=ELSE" .. string.char(0) .. "FOO=bar" ..string.char(0) .. "NEXT_VAR=42" print(c:match("FOO=(.*)")) '                     
barNEXT_VAR=42
```
Luckily, it seems like `environ` always ends with a zero byte, so `s:match("STARTUP_ID=(.*)" .. string.char(0))` should do the trick.